src/compiler/parser.ts

```diff
@@ -4640,9 +4647,11 @@ namespace ts {
                 //            ^^token; leftOperand = b. Return b ** c to the caller as a rightOperand
                 //      a ** b - c
                 //             ^token; leftOperand = b. Return b to the caller as a rightOperand
-                const consumeCurrentOperator = token() === SyntaxKind.AsteriskAsteriskToken ?
-                    newPrecedence >= precedence :
-                    newPrecedence > precedence;
+                const isPrivateIdentifierInInExpression = token() === SyntaxKind.InKeyword && isPrivateIdentifier(leftOperand);
+                const consumeCurrentOperator = isPrivateIdentifierInInExpression ||

```

Comment:
I think it would be better to parse a private identifier unconditionally and add a grammar error in checkPrivateIdentifier if the identifier's parent is incorrect. This way, it becomes a lot more complicated to convince myself that binary expression parsing is corrrect.

This also means that

1. The change in parseBinaryExpressionOrHigher is not needed.
2. checkExpression will need to call checkPrivateIdentifier to make sure that all incorrect uses of private identifiers get errors.

Reply:
You are 100% on the money with this. Turns out the parsing was incorrect, and so is the parsing in v8 (about to raise a ticket).

I'm changing things to the way you said and it works much better.

Reply:
Parser logic updated now. At first I thought going with this approach had broken the tests but turned out the tests were wrong, and this new approach leads to the correct result.

When I wrote the tests for this back in Febuary there were not implementations to try, and babel did not get the parsing right (they have now addressed these issues). So I unfortunately developed the wrong intuition for how this syntax should parse when nested within other expressions.

Did spot an issue in V8 when trying to compare to what they do: https://bugs.chromium.org/p/v8/issues/detail?id=12259